File: Classical Algorithm/QuickSort.java
package ymy1248.alg;

public class QuickSort{
    public static void sort(Comparable[] comps){
        sort(comps, 0, comps.length - 1);
    }

    private static void sort(Comparable[] comps, int lo, int hi){
        if(hi <= lo) return;
        int j = partition(comps, lo, hi);
        sort(comps, lo, j-1);
        sort(comps, j+1, hi);
    }

    private static int partition(Comparable[] comps, int lo, int hi){
        int i = lo;
        int j = hi + 1;
        Comparable p = comps[lo];
        while(true){
            while(comps[++i].compareTo(p) < 0)
                if(i == hi) break;
            while(comps[--j].compareTo(p) > 0)
                if(j == lo) break;
            if(i >= j) break;
            swap(comps, i, j);
        }
        swap(comps, lo, j);
        return j;
    }

    private static void swap(Comparable[] comps, int i, int j){
        Comparable reg = comps[i];
        comps[i] = comps[j];
        comps[j] = reg;
    }
}

File: C++_Algorithm_DataStructure/QuickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

template <typename T>
class QuickSort {
public:
    static void sort(vector<T> &arr) {
        sort(arr, 0, arr.size() - 1);
    }

private:
    static void sort(vector<T> &arr, size_t lo, size_t hi) {
        if (hi <= lo) return;
        size_t j = partition(arr, lo, hi);
        sort(arr, lo, j - 1);
        sort(arr, j + 1, hi);
    }

    static size_t partition(vector<T> &arr, size_t lo, size_t hi) {
        T piv = arr[lo];
        size_t i = lo;
        size_t j = hi + 1;
        
        while (true) {
            while (arr[++i] < piv) {
                if(i == hi) break;
            }
            while (arr[--j] > piv) {
                if(j == lo) break;
            }
            if (i >= j) {
                break;
            }
            T reg = arr[i];
            arr[i] = arr[j];
            arr[j] = reg;
        }
        T reg = arr[j];
        arr[j] = arr[lo];
        arr[lo] = reg;
        return j;
    }
};

int main() {
    vector<int> a {10,3,25,4321,5,53,21,432,143,21,65,632,32};
    QuickSort<int>::sort(a);
    
    for (size_t i = 0; i < a.size(); ++i) {
        cout << a[i] << " ";
    }
    cout << endl;
}
